### PR TITLE
complete FileWhitelist and NetworkMonitor

### PR DESCRIPTION
**Pull request by Gusabary**

### CODE REVIEW

**Comment by Gusabary** [comment]
basically finish

File: src/app/BasicRule.cpp [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+                    seccomp_rule_add(*this->ctxp, SCMP_ACT_TRACE(SM_EVM_BASIC_BASE + offset), rule.sysnum, 6, 
+                        rule.specs[0], rule.specs[1], rule.specs[2], rule.specs[3], rule.specs[4], rule.specs[5]);
+                    break;
+            }
+        }
+        offset++;
+    }
+}
+
+bool BasicRule::check(const long eventMsg, const user_regs_struct &regs, const int tid) {
```
Comment (Veiasai) [inline src/app/BasicRule.cpp:93]: this function is too lang. please split it into several small slices.

File: src/app/macro.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#pragma once
+
+// event message macro
```
Comment (Veiasai) [inline src/app/macro.h:5]: I dislike being abused by macro.

File: src/app/RuleManager.cpp [hunk 3]
```diff
@@ -1,50 +1,30 @@
...
+#include "NetworkMonitor.h"
 
 namespace SAIL { namespace rule {
 
-RuleManager::RuleManager(const std::string &configPath) {
-    this->ctx = seccomp_init(SCMP_ACT_ALLOW);
-
+RuleManager::RuleManager(const std::string &configPath, const std::shared_ptr<util::Utils> &up) : up(up) {
+    this->ctxp.reset(new scmp_filter_ctx(seccomp_init(SCMP_ACT_ALLOW)));
     const YAML::Node config = YAML::LoadFile(configPath);
```
Comment (Veiasai) [inline src/app/RuleManager.cpp]: configPath smells bad.
this constructor can't be mocked.
For example, in unit test, I wanna to build this object without a real file.

File: src/app/BasicRule.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        ptrace_arg_cmp(const unsigned int paraIndex, const std::string &action, const std::string &strValue) : 
+            paraIndex(paraIndex), action(action), strValue(strValue) {}
+    };
+
+    struct Rule {
+        int sysnum;
+        int id;
+        std::vector<struct scmp_arg_cmp> specs;
+        std::vector<struct ptrace_arg_cmp> pSpecs;
+        bool needExtraCheck; // if action is matchBytes or matchRe, which seccomp cannot handle, need extra check with ptrace
```
Comment (Veiasai) [inline src/app/BasicRule.h:32]: extraCheck is useful but not good enough.
the range of "extra" is not presented.

You should define the supported action explicitly.

**Review by Veiasai — CHANGES_REQUESTED** [review]
Good job.

File: src/app/macro.h (hunk 2 above)
Comment (Gusabary) [reply src/app/macro.h:5]: now wrap it with namespace

File: src/app/RuleManager.cpp (hunk 3 above)
Comment (Gusabary) [reply src/app/RuleManager.cpp]: fixed

File: src/app/BasicRule.cpp (hunk 1 above)
Comment (Gusabary) [reply src/app/BasicRule.cpp:93]: fixed

**Review by Veiasai — APPROVED** [review]

**Pull request merged**